hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/TransferExecutor.java

```diff
@@ -130,17 +144,65 @@ protected void executeCryptoTransfer(
             CryptoTransferStreamBuilder recordBuilder,
             boolean skipCustomFee) {
         final var topLevelPayer = context.payer();
-        // Use the op with replaced aliases in further steps
         transferContext.validateHbarAllowances();
 
         // Replace all aliases in the transaction body with its account ids
         final var replacedOp = ensureAndReplaceAliasesInOp(txn, transferContext, context, validator);
+
         // Use the op with replaced aliases in further steps
-        final var steps = decomposeIntoSteps(replacedOp, topLevelPayer, transferContext, skipCustomFee);
+        final List<TransferStep> steps = new ArrayList<>();
+        steps.add(new AssociateTokenRecipientsStep(replacedOp));
+        final var customFeeStep = new CustomFeeAssessmentStep(replacedOp);
+
+        List<CryptoTransferTransactionBody> txns = List.of(replacedOp);
+        if (!skipCustomFee) {
+            txns = customFeeStep.assessCustomFees(transferContext);
+        }
+
+        final var hasHooks = hasHooks(replacedOp);
+        HookCalls hookCalls = null;
+
+        if (hasHooks) {
+            final var assessedFeesWithPayerDebits = transferContext.getAssessedFeesWithPayerDebits();
+            // Extract the HookCalls from the transaction bodies after custom fee assessment
+            hookCalls =
+                    hookCallFactory.from(transferContext.getHandleContext(), replacedOp, assessedFeesWithPayerDebits);
+            dispatchHookCalls(
+                    hookCalls.context(),
+                    hookCalls.preOnlyHooks(),
+                    transferContext.getHandleContext(),
+                    HooksABI.FN_ALLOW);
+            dispatchHookCalls(
+                    hookCalls.context(),
+                    hookCalls.prePostHooks(),
+                    transferContext.getHandleContext(),
+                    HooksABI.FN_ALLOW_PRE);
+        }
+
+        for (final var t : txns) {
+            steps.add(new AssociateTokenRecipientsStep(t));
```

Comment:
Is there any reason to construct this `steps` list completely and _then_ iterate through it?

Why not just,
```
for (final var t : txns) {
    new AssociateTokenRecipientsStep(t).doIn(transferContext);
    new AdjustHbarChangesStep(t, topLevelPayer).doIn(transferContext);
    // Etc etc
```

Reply:
yeah, no reason. Made changes